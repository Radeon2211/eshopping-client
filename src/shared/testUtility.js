// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const historyPageNum = (pageNumber) => ({
  listen: jest.fn(),
  location: { search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push: jest.fn(),
});

export const propsPagination = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  maxQuantityPerPage: 2,
});
