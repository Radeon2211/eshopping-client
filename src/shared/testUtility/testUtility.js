// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';
import { v4 as uuidv4 } from 'uuid';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from '../../store/reducers/rootReducer';
import { productConditions } from '../constants';

export const defaultDeliveryAddress = {
  firstName: 'firstName',
  lastName: 'lastName',
  street: 'street',
  zipCode: 'zipCode',
  city: 'city',
  country: 'country',
  phone: 'phone',
};

export const defaultUserProfile = {
  ...defaultDeliveryAddress,
  _id: 'userId',
  username: 'username',
  email: 'email@domain.com',
  status: 'active',
  contacts: {
    email: true,
    phone: true,
  },
  createdAt: '2021-02-10T19:10:38.872Z',
};

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const createHistoryPageNumber = (pageNumber = 1, push = jest.fn()) => ({
  listen: jest.fn(),
  location: { pathname: '/products', search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push,
});

export const createPaginationProps = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  quantityPerPage: 2,
});

export const createProductItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
  quantitySold = 0,
  buyerQuantity = 0,
  description = '',
  condition = productConditions.NEW,
) => ({
  _id: productId || uuidv4(),
  name,
  price,
  quantity,
  quantitySold,
  buyerQuantity,
  description,
  condition,
  photo,
  seller: {
    username: sellerUsername,
  },
  __v: 0,
  createdAt: '2021-02-10T19:10:38.872Z',
  updatedAt: '2021-02-10T19:10:38.872Z',
});

export const createCartItem = (
  sellerUsername = 'username',
  quantity = 1,
  productId,
  price = 2,
  name = 'product name',
  productQuantity = 2,
  photo = false,
) => {
  const product = createProductItem(
    productId || uuidv4(),
    sellerUsername,
    productQuantity,
    price,
    name,
    photo,
  );

  return {
    _id: uuidv4(),
    quantity,
    product,
  };
};

export const createTransactionAndOrderProdItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
) => ({
  _id: productId || uuidv4(),
  name,
  photo,
  price,
  quantity,
  seller: {
    username: sellerUsername,
  },
});

export const createOrder = (
  products = [createTransactionAndOrderProdItem()],
  orderId = uuidv4(),
  sellerUsername,
  buyerUsername,
  overallPrice,
  createdAt,
  sellerEmail,
  sellerPhone,
) => {
  const seller = sellerUsername
    ? {
        username: sellerUsername,
        email: sellerEmail,
        phone: sellerPhone,
      }
    : null;

  const buyer = buyerUsername
    ? {
        username: buyerUsername,
      }
    : null;

  return {
    _id: orderId,
    seller,
    buyer,
    products,
    overallPrice,
    deliveryAddress: defaultDeliveryAddress,
    createdAt,
    updatedAt: createdAt,
    __v: 0,
  };
};

export const testStore = (initialAuth = {}, initialProduct = {}, initialUI = {}) => {
  const createStoreWithMiddleware = applyMiddleware(thunk)(createStore);
  const defaultState = createStoreWithMiddleware(rootReducer).getState();

  const finalState = {
    auth: {
      ...defaultState.auth,
      ...initialAuth,
    },
    product: {
      ...defaultState.product,
      ...initialProduct,
    },
    ui: {
      ...defaultState.ui,
      ...initialUI,
    },
  };
  const finalStore = createStoreWithMiddleware(rootReducer, finalState);

  return {
    store: finalStore,
    initialState: finalStore.getState(),
  };
};

export const checkReqMethodAndURL = (moxiosInstance, method, url, reqNumber = 0) => {
  if (reqNumber) {
    const requests = moxiosInstance.requests.__items;
    const prevRequest = requests[reqNumber - 1];
    return method === prevRequest.config.method && url === prevRequest.config.url;
  }
  const lastRequest = moxiosInstance.requests.mostRecent();
  return method === lastRequest.config.method && url === lastRequest.config.url;
};

export const createExpectedState = (
  initialState,
  expectedAuth = {},
  expectedProduct = {},
  expectedUI = {},
) => {
  const newState = {
    auth: {
      ...initialState.auth,
      ...expectedAuth,
    },
    product: {
      ...initialState.product,
      ...expectedProduct,
    },
    ui: {
      ...initialState.ui,
      ...expectedUI,
    },
  };

  return newState;
};

export const setUpStoreWithDefaultProfile = (auth = {}, product = {}, ui = {}) => {
  return testStore(
    {
      profile: defaultUserProfile,
      deliveryAddress: defaultDeliveryAddress,
      cart: [],
      ...auth,
    },
    product,
    ui,
  );
};
