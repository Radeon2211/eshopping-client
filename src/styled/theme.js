const theme = {
  colors: {
    light1: '#FAFAFA',
    light1Transparent: '#FAFAFA88',
    light2: '#DADFE1',
    light3: '#BFBFBF',
    light4: '#666',
    blue: '#1E8BC3',
    blueLight: '#22A7F0',
    blueDark: '#67809F',
    green: '#00A790',
    greenDark: '#009680',
    red: '#F22613',
  },
  shadows: {
    level1: '0 0 2px rgba(0, 0, 0, .1)',
    level2: '0 0 2px rgba(0, 0, 0, .3)',
    level3: '0 0 3px rgba(0, 0, 0, .3)',
    level4: '0 0 5px rgba(0, 0, 0, .3)',
    level5: '0 0 8px rgba(0, 0, 0, .3)',
  },
  fonts: {
    logo: 'Lobster, cursive',
    heading: 'Ubuntu, sans-serif',
    text: 'Lato, sans-serif',
  },
  spacings: {
    level1: '0.8rem',
    level2: '1.2rem',
    level3: '2.4rem',
    level4: '3.6rem',
    level5: '4.8rem',
  },
  zIndexes: {
    level1: 10,
    level2: 100,
    level3: 1000,
    level4: 10000,
  },
};

export default theme;
