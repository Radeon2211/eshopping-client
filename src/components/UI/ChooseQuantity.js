import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import NumberInput from './NumberInput';
import MyIcon from './MyIcon';
import { ReactComponent as MinusIcon } from '../../images/SVG/minus.svg';
import { ReactComponent as PlusIcon } from '../../images/SVG/plus.svg';
import themeStyles from '../../styled/theme';

const SC = {};
SC.Wrapper = styled.div`
  align-items: stretch;
  display: flex;

  & .button {
    background-color: transparent;
    border: 1px solid ${({ theme }) => theme.colors.light4};
    border-radius: 1px;
    cursor: pointer;
    outline: none;
    padding: 0 ${({ theme }) => theme.spacings.level2};

    &[disabled] {
      cursor: default;
    }

    &.minus {
      border-right: 0;
    }

    &.plus {
      border-left: 0;
    }
  }
`;

const ChooseQuantity = (props) => {
  const { name, value, maxQuantity, incremented, decremented, changed, blured } = props;

  return (
    <SC.Wrapper>
      <button type="button" className="button minus" disabled={value <= 1} onClick={decremented}>
        <MyIcon size="small" color={value <= 1 ? themeStyles.colors.light2 : ''}>
          <MinusIcon />
        </MyIcon>
      </button>
      <NumberInput
        name={name}
        size="small"
        value={value}
        changed={changed}
        blured={blured}
        focusable={false}
      />
      <button
        type="button"
        className="button plus"
        disabled={value >= maxQuantity}
        onClick={incremented}
      >
        <MyIcon size="small" color={value >= maxQuantity ? themeStyles.colors.light2 : ''}>
          <PlusIcon />
        </MyIcon>
      </button>
    </SC.Wrapper>
  );
};

ChooseQuantity.propTypes = {
  name: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
  maxQuantity: PropTypes.number.isRequired,
  incremented: PropTypes.func.isRequired,
  decremented: PropTypes.func.isRequired,
  changed: PropTypes.func.isRequired,
  blured: PropTypes.func.isRequired,
};

export default ChooseQuantity;
