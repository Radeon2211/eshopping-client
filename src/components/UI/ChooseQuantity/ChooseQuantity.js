import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import NumberInput from '../NumberInput/NumberInput';
import MyIcon from '../MyIcon';
import { ReactComponent as MinusIcon } from '../../../images/icons/minus.svg';
import { ReactComponent as PlusIcon } from '../../../images/icons/plus.svg';
import themeStyles from '../../../styled/theme';

const SC = {};
SC.Wrapper = styled.div`
  align-items: stretch;
  display: flex;

  & .button {
    background-color: transparent;
    border: 1px solid ${({ theme }) => theme.colors.light4};
    border-radius: 1px;
    cursor: pointer;
    outline: none;
    padding: 0 ${({ theme }) => theme.spacings.level2};

    &[disabled] {
      cursor: default;
    }

    &.minus {
      border-right: 0;
    }

    &.plus {
      border-left: 0;
    }
  }
`;

const ChooseQuantity = (props) => {
  const { name, value, maxQuantity, incremented, decremented, changed, blured, focused } = props;

  return (
    <SC.Wrapper>
      <button
        type="button"
        className="button minus"
        disabled={value <= 1}
        onClick={decremented}
        data-testid="ChooseQuantity-minus-btn"
        aria-label="Decrease product quantity"
      >
        <MyIcon
          size="small"
          color={value <= 1 ? themeStyles.colors.light2 : ''}
          data-testid="ChooseQuantity-minus-icon"
        >
          <MinusIcon />
        </MyIcon>
      </button>
      <NumberInput
        name={name}
        size="small"
        value={value}
        changed={changed}
        blured={blured}
        focused={focused}
      />
      <button
        type="button"
        className="button plus"
        disabled={value >= maxQuantity}
        onClick={incremented}
        data-testid="ChooseQuantity-plus-btn"
        aria-label="Increase product quantity"
      >
        <MyIcon
          size="small"
          color={value >= maxQuantity ? themeStyles.colors.light2 : ''}
          data-testid="ChooseQuantity-plus-icon"
        >
          <PlusIcon />
        </MyIcon>
      </button>
    </SC.Wrapper>
  );
};

ChooseQuantity.defaultProps = {
  focused: () => {},
};

ChooseQuantity.propTypes = {
  name: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
  maxQuantity: PropTypes.number.isRequired,
  incremented: PropTypes.func.isRequired,
  decremented: PropTypes.func.isRequired,
  changed: PropTypes.func.isRequired,
  blured: PropTypes.func.isRequired,
  focused: PropTypes.func,
};

export default ChooseQuantity;
