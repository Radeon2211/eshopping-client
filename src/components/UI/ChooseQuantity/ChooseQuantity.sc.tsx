import styled from 'styled-components';

export const Wrapper = styled.div`
  align-items: stretch;
  display: flex;

  & .button {
    background-color: transparent;
    border: 1px solid ${({ theme }) => theme.colors.light4};
    border-radius: 1px;
    cursor: pointer;
    outline: none;
    padding: 0 ${({ theme }) => theme.spacings.level2};

    &[disabled] {
      cursor: default;
    }

    &.minus {
      border-right: 0;
    }

    &.plus {
      border-left: 0;
    }
  }
`;
