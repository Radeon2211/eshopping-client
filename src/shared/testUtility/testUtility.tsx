/* eslint-disable import/no-extraneous-dependencies */
import { ReactElement } from 'react';
import { Router } from 'react-router-dom';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import { MockStoreEnhanced } from 'redux-mock-store';
import { fireEvent, waitFor, render } from '@testing-library/react';
import { v4 as uuidv4 } from 'uuid';
import moxios from 'moxios';
import rootReducer from '../../store/reducers/rootReducer';
import theme from '../../styled/theme';
import {
  CartItem,
  DeliveryAddress,
  Order,
  OrderProductItem,
  Product,
  Profile,
  ProfileStatus,
} from '../types/types';
import { ProductCondition } from '../types/enums';

export const defaultDeliveryAddress: DeliveryAddress = {
  firstName: 'firstName',
  lastName: 'lastName',
  street: 'street',
  zipCode: 'zipCode',
  city: 'city',
  country: 'country',
  phone: 'phone',
};

export const defaultUserProfile: Profile = {
  ...defaultDeliveryAddress,
  _id: 'userId',
  username: 'username',
  email: 'email@domain.com',
  status: ProfileStatus.ACTIVE,
  contacts: {
    email: true,
    phone: true,
  },
  createdAt: '2021-02-10T19:10:38.872Z',
};

export const createPaginationProps = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  quantityPerPage: 2,
});

export const createProductItem = (props: Partial<Product> = {}): Product => {
  return {
    _id: uuidv4(),
    quantity: 1,
    price: 2,
    name: 'product name',
    photo: false,
    quantitySold: 0,
    buyerQuantity: 0,
    description: '',
    condition: ProductCondition.NEW,
    seller: {
      username: 'username',
    },
    createdAt: '2021-02-10T19:10:38.872Z',
    updatedAt: '2021-02-10T19:10:38.872Z',
    __v: 0,
    ...props,
  };
};

export const createCartItem = (props = {}): CartItem => {
  const finalProps = {
    id: uuidv4(),
    sellerUsername: 'username',
    quantity: 1,
    productId: uuidv4(),
    price: 2,
    name: 'product name',
    productQuantity: 2,
    photo: false,
    ...props,
  };
  const product = createProductItem({
    _id: finalProps.productId,
    quantity: finalProps.productQuantity,
    price: finalProps.price,
    name: finalProps.name,
    photo: finalProps.photo,
    seller: {
      username: finalProps.sellerUsername,
    },
  });

  return {
    _id: finalProps.id,
    quantity: finalProps.quantity,
    product,
  };
};

export const createTransactionAndOrderProdItem = (
  props: Partial<OrderProductItem> = {},
): OrderProductItem => {
  return {
    _id: uuidv4(),
    quantity: 1,
    price: 2,
    name: 'product name',
    photo: false,
    seller: {
      username: 'username',
    },
    ...props,
  };
};

export const createOrder = (props: Partial<Order> = {}): Order => {
  return {
    _id: uuidv4(),
    overallPrice: 100,
    seller: null,
    buyer: null,
    products: [createTransactionAndOrderProdItem()],
    deliveryAddress: defaultDeliveryAddress,
    createdAt: '2021-02-10T19:10:38.872Z',
    updatedAt: '2021-02-10T19:10:38.872Z',
    __v: 0,
    ...props,
  };
};

export const testStore = (initialAuth = {}, initialProduct = {}, initialUI = {}) => {
  const createStoreWithMiddleware = applyMiddleware(thunk)(createStore);
  const defaultState = createStoreWithMiddleware(rootReducer).getState();

  const finalState = {
    auth: {
      ...defaultState.auth,
      ...initialAuth,
    },
    product: {
      ...(defaultState.product as object),
      ...initialProduct,
    },
    ui: {
      ...(defaultState.ui as object),
      ...initialUI,
    },
  };
  // @ts-expect-error TBF after redux upgrade
  const finalStore = createStoreWithMiddleware(rootReducer, finalState);

  return {
    store: finalStore,
    initialState: finalStore.getState(),
  };
};

type MoxiosRequestsItems = { config: { method: string; url: string } }[];
type MockedMoxiosInstance = {
  requests: {
    __items?: MoxiosRequestsItems;
    mostRecent?: () => { config: { method: string; url: string } };
  };
};
type DefaultMoxiosInstance = Omit<typeof moxios, 'requests'> & {
  requests: typeof moxios.requests & {
    __items?: MoxiosRequestsItems;
  };
};
export const checkReqMethodAndURL = (
  moxiosInstance: MockedMoxiosInstance | DefaultMoxiosInstance,
  method: string,
  url: string,
  reqNumber = 0,
) => {
  if (reqNumber) {
    const requests = moxiosInstance.requests.__items;
    const prevRequest = requests?.[reqNumber - 1];
    return prevRequest && method === prevRequest.config.method && url === prevRequest.config.url;
  }
  if (!moxiosInstance.requests.mostRecent) return false;
  const lastRequest = moxiosInstance.requests.mostRecent();
  return method === lastRequest.config.method && url === lastRequest.config.url;
};

export const createExpectedState = (
  initialState: {
    auth?: object;
    product?: object;
    ui?: object;
  },
  expectedAuth = {},
  expectedProduct = {},
  expectedUI = {},
) => {
  const newState = {
    auth: {
      ...initialState.auth,
      ...expectedAuth,
    },
    product: {
      ...initialState.product,
      ...expectedProduct,
    },
    ui: {
      ...initialState.ui,
      ...expectedUI,
    },
  };

  return newState;
};

export const setUpStoreWithDefaultProfile = (auth = {}, product = {}, ui = {}) => {
  return testStore(
    {
      profile: defaultUserProfile,
      deliveryAddress: defaultDeliveryAddress,
      cart: [],
      ...auth,
    },
    product,
    ui,
  );
};

export const clickAtSubmitButton = async (container: Element) => {
  await waitFor(() => {
    const submitBtn = container.querySelector('button[type="submit"]');
    if (submitBtn) {
      fireEvent.click(submitBtn);
    }
  });
};

export const testRouterPushCall = (
  pushFn: ReturnType<typeof jest.fn>,
  callIdx: number,
  expectedPathname: string,
  expectedSearch = '',
) => {
  expect(pushFn.mock.calls[callIdx][0].pathname).toEqual(expectedPathname);
  expect(pushFn.mock.calls[callIdx][0].search).toEqual(expectedSearch);
};

type RenderAppPartOptions = {
  push?: ReturnType<typeof jest.fn>;
  pathname?: string;
  search?: string;
  withoutThemeProvider?: boolean;
  withoutRouter?: boolean;
  store?: MockStoreEnhanced<object, object>;
  onlyReturnWrappedElement?: boolean;
};

export const renderAppPart = (element: ReactElement, options: RenderAppPartOptions = {}) => {
  const navigator = {
    createHref: jest.fn(),
    push: options.push || jest.fn(),
    go: jest.fn(),
    replace: jest.fn(),
  };

  const location = {
    pathname: options.pathname || '/',
    search: options.search || '',
  };

  let content = element;
  if (!options.withoutThemeProvider) {
    content = <ThemeProvider theme={theme}>{element}</ThemeProvider>;
  }
  if (options.store) {
    content = <Provider store={options.store}>{content}</Provider>;
  }
  if (!options.withoutRouter) {
    content = (
      <Router location={location} navigator={navigator}>
        {content}
      </Router>
    );
  }

  return options.onlyReturnWrappedElement ? content : render(content);
};
