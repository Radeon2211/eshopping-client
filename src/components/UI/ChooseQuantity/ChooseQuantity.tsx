import PropTypes from 'prop-types';
import * as SC from './ChooseQuantity.sc';
import NumberInput from '../NumberInput/NumberInput';
import MyIcon from '../MyIcon';
import { ReactComponent as MinusIcon } from '../../../images/icons/minus.svg';
import { ReactComponent as PlusIcon } from '../../../images/icons/plus.svg';
import themeStyles from '../../../styled/theme';

export interface ChooseQuantityProps {
  name: string;
  value: number;
  maxQuantity: number;
  incremented: () => void;
  decremented: () => void;
  changed: () => void;
  blured: () => void;
  focused?: () => void;
}

export default function ChooseQuantity(props: ChooseQuantityProps) {
  const { name, value, maxQuantity, incremented, decremented, changed, blured, focused } = props;

  return (
    <SC.Wrapper>
      <button
        type="button"
        className="button minus"
        disabled={value <= 1}
        onClick={decremented}
        data-testid="ChooseQuantity-minus-btn"
        aria-label="Decrease product quantity"
      >
        <MyIcon
          $size="small"
          $color={value <= 1 ? themeStyles.colors.light2 : ''}
          data-testid="ChooseQuantity-minus-icon"
        >
          <MinusIcon />
        </MyIcon>
      </button>
      <NumberInput
        name={name}
        size="small"
        value={value}
        changed={changed}
        blured={blured}
        focused={focused}
      />
      <button
        type="button"
        className="button plus"
        disabled={value >= maxQuantity}
        onClick={incremented}
        data-testid="ChooseQuantity-plus-btn"
        aria-label="Increase product quantity"
      >
        <MyIcon
          $size="small"
          $color={value >= maxQuantity ? themeStyles.colors.light2 : ''}
          data-testid="ChooseQuantity-plus-icon"
        >
          <PlusIcon />
        </MyIcon>
      </button>
    </SC.Wrapper>
  );
}

ChooseQuantity.defaultProps = {
  focused: () => {},
};

ChooseQuantity.propTypes = {
  name: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
  maxQuantity: PropTypes.number.isRequired,
  incremented: PropTypes.func.isRequired,
  decremented: PropTypes.func.isRequired,
  changed: PropTypes.func.isRequired,
  blured: PropTypes.func.isRequired,
  focused: PropTypes.func,
};
