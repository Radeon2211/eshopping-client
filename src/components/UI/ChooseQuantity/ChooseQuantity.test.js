import React from 'react';
import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { ThemeProvider } from 'styled-components';
import ChooseQuantity from './ChooseQuantity';
import { checkProps } from '../../../shared/testUtility';
import theme from '../../../styled/theme';

const defaultProps = {
  name: 'quantity',
  value: 1,
  maxQuantity: 5,
  incremented: jest.fn(),
  decremented: jest.fn(),
  changed: jest.fn(),
  blured: jest.fn(),
};

const setUp = (props = {}) => {
  const finalProps = props
    ? {
        ...defaultProps,
        ...props,
      }
    : defaultProps;

  return render(
    <ThemeProvider theme={theme}>
      <ChooseQuantity {...finalProps} />
    </ThemeProvider>,
  );
};

afterEach(cleanup);

describe('<ChooseQuantity />', () => {
  describe('Check prop types', () => {
    it('Should NOT throw a warning', () => {
      expect(checkProps(ChooseQuantity, defaultProps)).toBeUndefined();
    });

    it('Should throw a warning', () => {
      expect(checkProps(ChooseQuantity, {})).not.toBe(null);
    });
  });

  describe('Check how renders', () => {
    it('Should render everything correctly', () => {
      const { asFragment } = setUp();
      expect(asFragment()).toMatchSnapshot();
    });

    it('Should minus button be disabled but plus button NOT', () => {
      setUp();

      expect(screen.getByTestId('ChooseQuantity-minus-icon')).toHaveAttribute(
        'color',
        theme.colors.light2,
      );
      expect(screen.getByTestId('ChooseQuantity-plus-icon')).toHaveAttribute('color', '');

      expect(screen.getByTestId('ChooseQuantity-minus-btn')).toBeDisabled();
      expect(screen.getByTestId('ChooseQuantity-plus-btn')).not.toBeDisabled();
    });

    it('Should minus button NOT be disabled but plus button yes', () => {
      const props = {
        value: 5,
      };
      setUp(props);

      expect(screen.getByTestId('ChooseQuantity-minus-icon')).toHaveAttribute('color', '');
      expect(screen.getByTestId('ChooseQuantity-plus-icon')).toHaveAttribute(
        'color',
        theme.colors.light2,
      );

      expect(screen.getByTestId('ChooseQuantity-minus-btn')).not.toBeDisabled();
      expect(screen.getByTestId('ChooseQuantity-plus-btn')).toBeDisabled();
    });

    it('Should both buttons NOT be disabled', () => {
      const props = {
        value: 2,
      };
      setUp(props);

      expect(screen.getByTestId('ChooseQuantity-minus-icon')).toHaveAttribute('color', '');
      expect(screen.getByTestId('ChooseQuantity-plus-icon')).toHaveAttribute('color', '');

      expect(screen.getByTestId('ChooseQuantity-minus-btn')).not.toBeDisabled();
      expect(screen.getByTestId('ChooseQuantity-plus-btn')).not.toBeDisabled();
    });
  });

  describe('Check behaviour', () => {
    it('Should call changed', () => {
      const changedFn = jest.fn();
      const props = {
        changed: changedFn,
      };
      setUp(props);

      fireEvent.input(screen.getByTestId('NumberInput-quantity'), { target: { value: 5 } });
      expect(changedFn).toHaveBeenCalledTimes(1);
    });

    it('Should call incremented', () => {
      const incrementedFn = jest.fn();
      const props = {
        incremented: incrementedFn,
      };
      setUp(props);

      fireEvent.click(screen.getByTestId('ChooseQuantity-plus-btn'));
      expect(incrementedFn).toHaveBeenCalledTimes(1);
    });

    it('Should NOT call incremented if plus button is disabled', () => {
      const incrementedFn = jest.fn();
      const props = {
        value: 5,
        incremented: incrementedFn,
      };
      setUp(props);

      fireEvent.click(screen.getByTestId('ChooseQuantity-plus-btn'));
      expect(incrementedFn).not.toHaveBeenCalled();
    });

    it('Should call decremented', () => {
      const decrementedFn = jest.fn();
      const props = {
        decremented: decrementedFn,
        value: 2,
      };
      setUp(props);

      fireEvent.click(screen.getByTestId('ChooseQuantity-minus-btn'));
      expect(decrementedFn).toHaveBeenCalledTimes(1);
    });

    it('Should NOT call decremented if minus button is disabled', () => {
      const decrementedFn = jest.fn();
      const props = {
        decremented: decrementedFn,
      };
      setUp(props);

      fireEvent.click(screen.getByTestId('ChooseQuantity-minus-btn'));
      expect(decrementedFn).not.toHaveBeenCalled();
    });

    it('Should call blured and focused', () => {
      const bluredFn = jest.fn();
      const focusedFn = jest.fn();
      const props = {
        blured: bluredFn,
        focused: focusedFn,
      };
      setUp(props);

      const input = screen.getByTestId('NumberInput-quantity');
      fireEvent.focus(input);
      expect(focusedFn).toHaveBeenCalledTimes(1);
      fireEvent.blur(input);
      expect(bluredFn).toHaveBeenCalledTimes(1);
    });
  });
});
