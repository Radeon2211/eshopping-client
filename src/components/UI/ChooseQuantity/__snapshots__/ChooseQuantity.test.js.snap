// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChooseQuantity /> Check how renders Should render everything correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-AxiKw llMUIM"
  >
    <button
      class="button minus"
      data-testid="ChooseQuantity-minus-btn"
      disabled=""
      type="button"
    >
      <div
        class="sc-AxirZ hLmnqv"
        color="#DADFE1"
        data-testid="ChooseQuantity-minus-icon"
      >
        <svg>
          minus.svg
        </svg>
      </div>
    </button>
    <input
      class="sc-AxjAm gofkAj"
      data-testid="NumberInput-quantity"
      name="quantity"
      type="number"
      value="1"
    />
    <button
      class="button plus"
      data-testid="ChooseQuantity-plus-btn"
      type="button"
    >
      <div
        class="sc-AxirZ csXoMP"
        color=""
        data-testid="ChooseQuantity-plus-icon"
      >
        <svg>
          plus.svg
        </svg>
      </div>
    </button>
  </div>
</DocumentFragment>
`;
